src/d3a/models/strategy/settlement/strategy.py

```diff
@@ -0,0 +1,215 @@
+        self.bid_updater = SettlementTemplateStrategyBidUpdater(
+                initial_rate=initial_buying_rate,
+                final_rate=final_buying_rate,
+                fit_to_limit=True,
+                energy_rate_change_per_update=None,
+                update_interval=duration(minutes=self._update_interval),
+                rate_limit_object=min)
+
+        self.offer_updater = SettlementTemplateStrategyOfferUpdater(
+                initial_rate=initial_selling_rate,
+                final_rate=final_selling_rate,
+                fit_to_limit=True,
+                energy_rate_change_per_update=None,
+                update_interval=duration(minutes=self._update_interval),
+                rate_limit_object=max)
+
+    def event_market_cycle(self, strategy: BaseStrategy) -> None:
+        """
+        Should be called by the event_market_cycle of the asset strategy class, posts
+        settlement bids and offers on markets that do not have posted bids and offers yet
+        Args:
+            strategy: Strategy object of the asset
+
+        Returns: None
+
+        """
+        self.bid_updater.update_and_populate_price_settings(strategy.area)
+        self.offer_updater.update_and_populate_price_settings(strategy.area)
+        for market in strategy.area.settlement_markets.values():
+            energy_deviation_kWh = strategy.state.get_unsettled_deviation_kWh(market.time_slot)
+
+            if strategy.state.can_post_settlement_bid(market.time_slot):
+                try:
+                    strategy.post_first_bid(
+                        market, energy_deviation_kWh * 1000.0,
+                        self.bid_updater.initial_rate[market.time_slot]
+                    )
+                except MarketException:
+                    pass
+
+            if strategy.state.can_post_settlement_offer(market.time_slot):
+                try:
+                    strategy.post_first_offer(
+                        market, energy_deviation_kWh,
+                        self.offer_updater.initial_rate[market.time_slot]
+                    )
+                except MarketException:
+                    pass
+
+    def event_tick(self, strategy: BaseStrategy) -> None:
+        """
+        Update posted settlement bids and offers on market tick
+        Args:
+            strategy: Strategy object of the asset
+
+        Returns: None
+
+        """
+        self.bid_updater.increment_update_counter_all_markets(strategy)
+        self.offer_updater.increment_update_counter_all_markets(strategy)
+        for market in strategy.area.settlement_markets.values():
+            self.bid_updater.update(market, strategy)
+            self.offer_updater.update(market, strategy)
+
+    def _get_settlement_market_by_id(
+            self, strategy: BaseStrategy, market_id: str) -> Optional["Market"]:
+        markets = [m for m in strategy.area.settlement_markets.values() if m.id == market_id]
+        if not markets:
```

Comment:
```suggestion
        if not markets or len(markets) > 1:
```

As there shouldn't be more than one market with an id

Reply:
I am changing this to an assertion instead. The suggestion will hid the problem by returning None, while we should be aware that the settlement market list is corrupt in this case.